quickwit/quickwit-ingest/src/ingest_v2/shard_table.rs

```diff
@@ -24,46 +24,83 @@ use quickwit_proto::ingest::Shard;
 use quickwit_proto::types::SourceId;
 use quickwit_proto::IndexId;
 
-/// A set of open shards for a given index and source.
 #[derive(Debug, Default)]
-pub(super) struct ShardTableEntry {
+struct ShardTableEntryInfo {
     shards: Vec<Shard>,
     round_robin_idx: AtomicUsize,
 }
 
+/// A set of open shards for a given index and source.
+#[derive(Debug)]
+pub(super) struct ShardTableEntry {
+    local_shards: ShardTableEntryInfo,
+    remote_shards: ShardTableEntryInfo,
+}
+
 impl ShardTableEntry {
     /// Creates a new entry and ensures that the shards are open and unique.
     ///
     /// # Panics
     ///
     /// Panics if `shards` is empty after filtering out closed shards and deduplicating by shard ID.
-    pub fn new(mut shards: Vec<Shard>) -> Self {
+    pub fn new(mut shards: Vec<Shard>, node_id: String) -> Self {
         shards.retain(|shard| shard.is_open());
         shards.sort_unstable_by_key(|shard| shard.shard_id);
         shards.dedup_by_key(|shard| shard.shard_id);
 
         assert!(!shards.is_empty(), "`shards` should not be empty");
 
-        Self {
-            shards,
+        let local_shards_vec = &mut shards;
```

Comment:
We try not to overuse iterators and functional combinators, but in this instance, they work nicely:

```rust
let (local_shards, remove_shards): (Vec<_>, Vec<_>) = a
    .into_iter()
    .filter(|shard| shard.is_open())
    .sort_unstable_by_key(|shard| shard.shard_id)
    .dedup_by_key(|shard| shard.shard_id)
    .partition(|shard| shard.leader_id == self.self_node_id)
    .collect();

assert!(local_shards.len() + remote_shards.len() > 0 , "`shards` should not be empty");

Self {
    local_shards,
    local_round_robin_idx: AtomicUsize::default()
    remote_shards,
    remote_round_robin_idx: AtomicUsize::default()
}
```

Reply:
sort_unstable_by_key could not work on a filter so I kept the old code and added the partition function alone.